File: crates/defi-cli/src/output.rs
use crate::commands::Cli;
use serde::Serialize;

pub struct OutputMode {
    pub json: bool,
    pub ndjson: bool,
    pub fields: Option<Vec<String>>,
}

impl OutputMode {
    pub fn from_cli(cli: &Cli) -> Self {
        Self {
            json: cli.json || cli.ndjson,
            ndjson: cli.ndjson,
            fields: cli
                .fields
                .as_ref()
                .map(|f| f.split(',').map(|s| s.trim().to_string()).collect()),
        }
    }

    pub fn print<T: Serialize>(&self, value: &T) -> Result<(), defi_core::error::DefiError> {
        if self.ndjson {
            let s = serde_json::to_string(value)
                .map_err(|e| defi_core::error::DefiError::Internal(e.to_string()))?;
            println!("{s}");
        } else if self.json {
            let mut json_val = serde_json::to_value(value)
                .map_err(|e| defi_core::error::DefiError::Internal(e.to_string()))?;

            if let Some(ref fields) = self.fields
                && let serde_json::Value::Object(ref map) = json_val
            {
                let filtered: serde_json::Map<String, serde_json::Value> = map
                    .iter()
                    .filter(|(k, _)| fields.contains(k))
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect();
                json_val = serde_json::Value::Object(filtered);
            }

            println!("{}", serde_json::to_string_pretty(&json_val).unwrap());
        } else {
            // Human-readable: try table format, fallback to pretty JSON
            let json_val = serde_json::to_value(value)
                .map_err(|e| defi_core::error::DefiError::Internal(e.to_string()))?;
            if let Some(table) = crate::table::render(&json_val) {
                println!("{}", table);
            } else {
                println!("{}", serde_json::to_string_pretty(&json_val).unwrap());
            }
        }
        Ok(())
    }
}

File: ts/packages/defi-cli/src/output.ts
import { jsonStringify } from "@hypurrquant/defi-core";
import { renderTable } from "./table.js";

export interface OutputMode {
  json: boolean;
  ndjson: boolean;
  fields?: string[];
}

export function parseOutputMode(opts: {
  json?: boolean;
  ndjson?: boolean;
  fields?: string;
}): OutputMode {
  return {
    json: !!(opts.json || opts.ndjson),
    ndjson: !!opts.ndjson,
    fields: opts.fields
      ? opts.fields.split(",").map((f) => f.trim())
      : undefined,
  };
}

export function formatOutput(value: unknown, mode: OutputMode): string {
  if (mode.ndjson) {
    return JSON.stringify(value, jsonReplacer);
  }

  if (mode.json) {
    let jsonVal = JSON.parse(jsonStringify(value));

    if (mode.fields && typeof jsonVal === "object" && jsonVal !== null && !Array.isArray(jsonVal)) {
      const filtered: Record<string, unknown> = {};
      for (const key of mode.fields) {
        if (key in jsonVal) filtered[key] = jsonVal[key];
      }
      jsonVal = filtered;
    }

    return JSON.stringify(jsonVal, null, 2);
  }

  // Human-readable: try table format, fallback to pretty JSON
  const jsonVal = JSON.parse(jsonStringify(value));
  const table = renderTable(jsonVal);
  if (table !== null) return table;
  return JSON.stringify(jsonVal, null, 2);
}

export function printOutput(value: unknown, mode: OutputMode): void {
  console.log(formatOutput(value, mode));
}

function jsonReplacer(_key: string, val: unknown): unknown {
  if (typeof val === "bigint") return "0x" + val.toString(16);
  return val;
}
